@isbm The test failures for the Ubuntu 16 builds look suspicious to me. I am not seeing those failures on the HEAD of the branch.

I am also seeing many of these stacktraces:
```
15:37:53 AttributeError: module 'zmq' has no attribute 'eventloop'
15:37:53 Traceback (most recent call last):
15:37:53   File "/tmp/salt-tests-tmpdir/scripts/salt-call", line 5, in <module>
15:37:53     salt_call()
15:37:53   File "/testing/salt/scripts.py", line 395, in salt_call
15:37:53     client.run()
15:37:53   File "/testing/salt/cli/call.py", line 47, in run
15:37:53     caller = salt.cli.caller.Caller.factory(self.config)
15:37:53   File "/testing/salt/cli/caller.py", line 79, in factory
15:37:53     return ZeroMQCaller(opts, **kwargs)
15:37:53   File "/testing/salt/cli/caller.py", line 291, in __init__
15:37:53     super(ZeroMQCaller, self).__init__(opts)
15:37:53   File "/testing/salt/cli/caller.py", line 102, in __init__
15:37:53     self.minion = salt.minion.SMinion(opts)
15:37:53   File "/testing/salt/minion.py", line 785, in __init__
15:37:53     io_loop = zmq.eventloop.ioloop.ZMQIOLoop()
```
Can you look into that?

--- a/salt/minion.py
+++ b/salt/minion.py
@@ -781,10 +781,7 @@ class SMinion(MinionBase):
         # Clean out the proc directory (default /var/cache/salt/minion/proc)
         if (self.opts.get('file_client', 'remote') == 'remote'
                 or self.opts.get('use_master_when_local', False)):
-            if self.opts['transport'] == 'zeromq' and zmq:
-                io_loop = zmq.eventloop.ioloop.ZMQIOLoop()
-            else:
-                io_loop = ZMQDefaultLoop.current()
+            io_loop = ZMQDefaultLoop.current()
             io_loop.run_sync(
                 lambda: self.eval_master(self.opts, failed=True)
             )

--- a/salt/transport/zeromq.py
+++ b/salt/transport/zeromq.py
@@ -514,14 +514,7 @@ class ZeroMQReqServerChannel(salt.transport.mixins.auth.AESReqServerMixin,
             # IPv6 sockets work for both IPv6 and IPv4 addresses
             self.clients.setsockopt(zmq.IPV4ONLY, 0)
         self.clients.setsockopt(zmq.BACKLOG, self.opts.get('zmq_backlog', 1000))
-        if HAS_ZMQ_MONITOR and self.opts['zmq_monitor']:
-            # Socket monitor shall be used the only for debug
-            # purposes so using threading doesn't look too bad here
-            import threading
-            self._monitor = ZeroMQSocketMonitor(self.clients)
-            t = threading.Thread(target=self._monitor.start_poll)
-            t.start()
-
+        self._start_zmq_monitor()
         self.workers = self.context.socket(zmq.DEALER)
 
         if self.opts.get('ipc_mode', '') == 'tcp':
@@ -583,6 +576,21 @@ class ZeroMQReqServerChannel(salt.transport.mixins.auth.AESReqServerMixin,
         salt.transport.mixins.auth.AESReqServerMixin.pre_fork(self, process_manager)
         process_manager.add_process(self.zmq_device)
 
+    def _start_zmq_monitor(self):
+        '''
+        Starts ZMQ monitor for debugging purposes.
+        :return:
+        '''
+        # Socket monitor shall be used the only for debug
+        # purposes so using threading doesn't look too bad here
+
+        if HAS_ZMQ_MONITOR and self.opts['zmq_monitor']:
+            log.debug('Starting ZMQ monitor')
+            import threading
+            self._w_monitor = ZeroMQSocketMonitor(self._socket)
+            threading.Thread(target=self._w_monitor.start_poll).start()
+            log.debug('ZMQ monitor has been started started')
+
     def post_fork(self, payload_handler, io_loop):
         '''
         After forking we need to create all of the local sockets to listen to the
@@ -597,13 +605,7 @@ class ZeroMQReqServerChannel(salt.transport.mixins.auth.AESReqServerMixin,
 
         self.context = zmq.Context(1)
         self._socket = self.context.socket(zmq.REP)
-        if HAS_ZMQ_MONITOR and self.opts['zmq_monitor']:
-            # Socket monitor shall be used the only for debug
-            # purposes so using threading doesn't look too bad here
-            import threading
-            self._w_monitor = ZeroMQSocketMonitor(self._socket)
-            t = threading.Thread(target=self._w_monitor.start_poll)
-            t.start()
+        self._start_zmq_monitor()
 
         if self.opts.get('ipc_mode', '') == 'tcp':
             self.w_uri = 'tcp://127.0.0.1:{0}'.format(

--- a/salt/transport/zeromq.py
+++ b/salt/transport/zeromq.py
@@ -814,27 +814,35 @@ class ZeroMQPubServerChannel(salt.transport.server.PubServerChannel):
                 # Catch and handle EINTR from when this process is sent
                 # SIGUSR1 gracefully so we don't choke and die horribly
                 try:
+                    log.trace('Getting data from puller %s', pull_uri)
                     package = pull_sock.recv()
                     unpacked_package = salt.payload.unpackage(package)
                     if six.PY3:
                         unpacked_package = salt.transport.frame.decode_embedded_strs(unpacked_package)
                     payload = unpacked_package['payload']
+                    log.trace('Accepted unpacked package from puller')
                     if self.opts['zmq_filtering']:
                         # if you have a specific topic list, use that
                         if 'topic_lst' in unpacked_package:
                             for topic in unpacked_package['topic_lst']:
+                                log.trace('Sending filtered data over publisher %s', pub_uri)
                                 # zmq filters are substring match, hash the topic
                                 # to avoid collisions
                                 htopic = hashlib.sha1(topic).hexdigest()
                                 pub_sock.send(htopic, flags=zmq.SNDMORE)
                                 pub_sock.send(payload)
+                                log.trace('Filtered data has been sent')
                                 # otherwise its a broadcast
                         else:
                             # TODO: constants file for "broadcast"
+                            log.trace('Sending broadcasted data over publisher %s', pub_uri)
                             pub_sock.send('broadcast', flags=zmq.SNDMORE)
                             pub_sock.send(payload)
+                            log.trace('Broadcasted data has been sent')
                     else:
+                        log.trace('Sending ZMQ-unfiltered data over publisher %s', pub_uri)
                         pub_sock.send(payload)
+                        log.trace('Unfiltered data has been sent')
                 except zmq.ZMQError as exc:
                     if exc.errno == errno.EINTR:
                         continue

